design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,336 @@
+command is executed from this directory - Command::new(...).current_dir(...) should do); if does not exist we could just create it
+- create a hidden Scarb.toml with a custom name (eg `.script_name.toml`) and override path to scarb toml using scarb flag;
+if a file does not exist let's just create it
+
+Option 1 is probably more in line with scarb ethos, but the option 2 would probably be more practical for us as it would allow
+for multiple Scarb.toml files in one directory, where multiple scripts could reside + Scarb.toml file present in cwd would not
+confuse sncast subcommands (eg declare also looks for Scarb.toml in cwd, which could lead to errors/user confusion).
+
+The deployment script could then be run like so:
+```bash
+$ sncast script /path/to/myscript.s.cairo
+```
+
+Log info should be outputted to stdout and optionally written to some log file. 
+We could have a flag to suppress outputting to stdout or to output with different formats (int, hex).
+
+By default, the scripts could run in some kind of "dry run mode", meaning all the transactions would be simulated with 
+specific rpc calls (like `starknet_simulateTransactions`). An interesting idea would be to research an approach similar 
+to cheatnet - fork a runner and test the scripts against it (to be assessed if it is feasible). We could also provide 
+a way to show a trace of said transactions if required. In later iterations, we could also provide a helper function
+that sets up the devnet to be used for dry running for users.
+
+The behaviour would be changed with `--broadcast` flag, that would actually execute needed transactions.
+
+### error handling
+If the transaction fails, we should put all the relevant info about it to the state file, output log information and stop
+script execution.
+If the script fails (for any reason that is not connected to transactions - eg rpc node down), we should output relevant
+log information.
+
+In case of failed transaction, in the "output" field in state file we could include an error message. After fixing the issue
+we should allow users to just re-run the script - all the previous (succeeded) transactions should not be replayed, the
+erroneous transaction should be retried and its output should be put into [state file](#example-state-file).
+
+### idempotency
+At the later stages we will want to have the script to be able to track the on chain state using a state file, which would 
+in turn allow the script to be called multiple times without any modifications, to ensure desired state. To achieve that
+we must ensure idempotency of calls to the network.
+
+The proposed solution would be to use an external json file, that holds hashes of transactions that were already performed.
+While executing a command we could first check this file to see if given transaction exists/succeeded, and based on this
+either perform and an action (make a transaction) or skip it and proceed to the next thing in script.
```

Comment:
to see if the transaction was already executed, the proposal is to check for the function name and its arguments (making it some kind of a function signature), and the hash is just its result. So, if I understand you correctly, the transaction would be replayed only if something before it changed, causing the arguments change. An example would be that we do a `call` and mark it to be re-done each script invokation (so eg do not use `skip_if_done`), then you're right and the tx would be re-done. But, this would have to be deliberate. Otherwise, the arguments should not change, so effectively the tx won't be replayed.

Does that address your concern?

Reply:
What if somebody wants to do another deploy (i.e. second system) with the same parameters? I think it might be a valid use case

Reply:
At the moment, there are a few options:
- one makes sure there is nothing prior to deploy that changes state (if one does not desire such things - eg would be an account creation), then a `--replay` flag can be used to create a second system
- one can modify the state file to exclude txs they want to do (this is veery janky, and not really a solution but rather a workaround)

I understand that those solutions may not be optimal, but I'm not sure how else we should approach it. We could provide another argument (something like `--replay-since` that would indicate from which tx one would want to redo the txs), but I'm not sure this is something that will we desired - at least for first iterations. If you have any ideas on that matter, please feel free to share.

Reply:
Wouldn't just having separate state files per system enough to solve this?